src/main/groovy/net/researchgate/release/GitAdapter.groovy

```diff
@@ -146,9 +151,30 @@ class GitAdapter extends BaseScmAdapter {
 
     @Override
     void revert() {
+        // Try to go back to the requireBranch (workingBranch is irrelevant here, a new GitAdapter instance has been created specifically during revert)
+        if (extension.git.requireBranch) {
+            exec(['git', 'checkout', extension.git.requireBranch], directory: workingDirectory, errorMessage: 'Error reverting changes made by the release plugin.')
+        }
+        // Revert changes on gradle.properties
         exec(['git', 'checkout', findPropertiesFile().name], directory: workingDirectory, errorMessage: 'Error reverting changes made by the release plugin.')
     }
 
+    @Override
+    void checkoutMergeToReleaseBranch() {
+        checkoutMerge(workingBranch, releaseBranch)
+    }
+
+    @Override
+    void checkoutMergeFromReleaseBranch() {
+        checkoutMerge(releaseBranch, workingBranch)
+    }
+
+    private checkoutMerge(String fromBranch, String toBranch) {
+        exec(['git', 'fetch'], directory: workingDirectory, errorPatterns: ['error: ', 'fatal: '])
+        exec(['git', 'checkout', toBranch], directory: workingDirectory, errorPatterns: ['error: ', 'fatal: '])
+        exec(['git', 'merge', '--no-ff', '--no-commit', '--strategy=recursive', '-Xtheirs', fromBranch], directory: workingDirectory, errorPatterns: ['error: ', 'fatal: '])

```

Comment:
Why do use no fast forward here and do some kind of automerge? Do you want to have commits in the release branch that are not in the master? And with the "-Xtheirs" auto merging we have some kind of magic in their that automatically resolves conflicts without letting you know. I mean if something changes in release and you changed it in master different for some reason it would just resolve it to the release and the master would not reflect what is actually used in the release. I think this is some kind of bad behavior. What do you think about letting the release fail if there are real conflicts?

Reply:
Hello, and thank you for your review. So some points need to be discussed ^^

First, sorry for the semicolon, Java reflex, I will remove them.

Second, about the "theirs" strategy, you are right: conflicts should not occur when merging master to release, except if someone messed up with the release branch, which should be handled manually. So in this case, standard strategy should be used, and the build should fail. In theory, merging back release to master should be done with "theirs" strategy, as we do not want code from release to be reused on master. But at this point the only modification taken from release is  gradle.properties, which declares a release version not a snapshot one: this modification will be immediatly overwritten to declare a new snapshot version.

So third, why not use fast-forward instead of this more complex full merge from master to release, then release to master? Because of the process behind the purpose to have separated branchs release and master. I started this PR because I have the same situation than people who asked for this functionality at the beginning: I have an integration system compiling specifically master for snapshots, and release for stable versions. It means that one of the most important constraint is that the code of a release should never be on the master branch (and the inverse also for snapshot code towards release branch). Otherwise snapshot code could be processed by with a release logic, and be deployed to a production environment for instance...

So what will happen if we let Git use fast forward when it can? This :

![image](https://user-images.githubusercontent.com/9728851/38776091-c0bdb52e-4091-11e8-88e4-4c0ec1c61be9.png)

Indeed master and release can always be fast forwarded to each other, as it is highly improbable that someone else pushes a commit during the gradle task execution. So history of both branchs are completely merged. In most cases, integration systems will behave correctly, as they do not execute on each commit by default, but on the last commit of a group pushed together to remote. However we cannot be sure of that. And because of history merged, we cannot state if a given commit is dedicated to release or master without analyzing the code or the commit messages. 

With my approach, it gives:

![image](https://user-images.githubusercontent.com/9728851/38776098-d94857ac-4091-11e8-8614-9ed0fd437ccb.png)

The structure is more complex, but history is trivial: each commit is either on one branch or on the other. Even if the integration system executes on every commit, each commit can be processed on the right logic without possible misleadings.

So I think it is what people are waiting for when we are talking about handling dedicated release and develop branchs.